File: sbr-backend/models/ServiceRequest.js
const mongoose = require('mongoose');

const LocationPointSchema = new mongoose.Schema({
  latitude: {
    type: Number,
    required: true
  },
  longitude: {
    type: Number,
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  }
}, { _id: false });

const ServiceRequestSchema = new mongoose.Schema(
  {
    customerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Please add a customer ID']
    },
    assignedAgentId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null
    },
    serviceType: {
      type: String,
      required: [true, 'Please specify the service type']
    },
    description: {
      type: String
    },
    customerAddress: {
      type: String,
      required: [true, 'Please add the customer address']
    },
    status: {
      type: String,
      enum: ['Pending', 'Assigned', 'Accepted', 'In Progress', 'Completed', 'Cancelled'],
      default: 'Pending'
    },
    createdBy: {
      type: String,
      enum: ['CUSTOMER', 'ADMIN'],
      default: 'CUSTOMER'
    },
    acceptedAt: {
      type: Date
    },
    completedAt: {
      type: Date
    },
    beforeImageUrl: {
      type: String
    },
    afterImageUrl: {
      type: String
    },
    paymentAmount: {
      type: Number,
      default: 0.0
    },
    paymentStatus: {
      type: String,
      enum: ['Pending', 'Paid'],
      default: 'Pending'
    },
    paymentMethod: {
      type: String
    },
    paymentTimestamp: {
      type: Date
    },
    locationPath: {
      type: [LocationPointSchema],
      default: []
    }
  },
  {
    timestamps: true
  }
);

module.exports = mongoose.model('ServiceRequest', ServiceRequestSchema);

File: sbr ios/Models/ServiceRequest.swift
import Foundation

struct LocationPoint: Codable {
    let latitude: Double
    let longitude: Double
    let timestamp: Date?
}

enum RequestStatus: String, Codable {
    case pending = "Pending"
    case assigned = "Assigned"
    case accepted = "Accepted"
    case inProgress = "In Progress"
    case completed = "Completed"
    case cancelled = "Cancelled"
}

struct ServiceRequest: Codable, Identifiable {
    let id: String
    let customerId: User?
    let assignedAgentId: User?
    let serviceType: String
    let description: String?
    let customerAddress: String
    let status: RequestStatus
    let createdBy: String
    let acceptedAt: String?
    let completedAt: String?
    let beforeImageUrl: String?
    let afterImageUrl: String?
    let paymentAmount: Double?
    let paymentStatus: String?
    let paymentMethod: String?
    let paymentTimestamp: String?
    let locationPath: [LocationPoint]?
    let createdAt: String
    let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case customerId, assignedAgentId, serviceType, description, customerAddress, status, createdBy, acceptedAt, completedAt, beforeImageUrl, afterImageUrl, paymentAmount, paymentStatus, paymentMethod, paymentTimestamp, locationPath, createdAt, updatedAt
    }
}
